REFACTOR.md

```diff
@@ -0,0 +1,199 @@
+
+###### Definite additions:
+
+- transformRow: Transforms each row as they come in, similar to a stream
+- transformResult: Transforms the entire response, before resolving the promise
+
+###### Likely additions: 
+
+- beforeFirst: Before first query is executed for a context, useful for transactions / connection variable preparation
+- beforeQuery: Before a query is executed
+- afterQuery: After a query is executed, before the result is provided
+- beforeEnd: Before a connection is released from a context & put back in the pool, optionally async
+
+
+##### Ability to use knex without a connection pool / with clustered pools
+
+Folks have asked for this before, and I think creating a convention for supplying a config option for acquiring raw driver connections / supplying custom pools will do the trick here.
+
+##### Missing features like modifying columns, etc.
+
+I know column modification is a huge missing feature, this is something that should be easier to add in as things are cleaned up.
+
+----
+
+###### Possible Additions: 
+
+- interceptQuery: Intercept a query and respond with a fake value, maybe useful when testing
+
+##### Standardized value escaping per dialect
+
+Currently the value escaping happens in a few different places, sometimes in the knex client, sometimes in the underlying DB library, sometimes in helper functions for cases like Oracle, etc. What we'd like to do here is have a standardized way of knowing what is escaped where.
+
+##### Better logging
+
+There should be a configurable logging creation, which will use console.log by default but allow users to supply their own log configuration for debugging. An API spec will be added here once we get there.
+
+#### Reach goals
+
+1. `asPartial`
+1. test refactor
+1. properly constructed prepared statements
+1. modularize things into a monorepo (lerna or similar)
+1. seeds / migrations improvements
+
+##### asPartial
+
+Discussed in a few tickets, something that would be useful, but would require quite a bit of internal refactoring in the query building chain. Will add more info if/when we get here.
+
+##### Test refactor
+
+Tests are a bit of a mess right now, difficult to follow, test across different drivers etc. This is a big task, which is why it's added as a reach goal - but it's something that should happen 
+
+##### Properly constructed prepared statements
+
+The difficulty with properly constructed prepared statements is the requirement that prepared statements execute on the same connection. The "context" addition should help in getting closer to this, but that still does not help with knowing which connection to checkout from the pool. Something to keep in mind.
+
+##### Modularizing
+
+For knex, this isn't necessarily a goal - most of the pieces of knex aren't really re-usable outside of knex, but it'd be good to design things more modularly for testing purposes, etc.
```

Comment:
The query builder being it's own module would help in cases like https://www.npmjs.com/package/cartodb-tools